proposals/massive_context_refactor.cpp

```diff
@@ -0,0 +1,381 @@
+typedef utensor::string OpName;
+typedef std::unordered_map<TName, Tensor*> TensorNamePtrMap;
+ 
+class TensorRecord {
+    public:
+        unsigned char ref_count = 0;
+        Tensor* t
+        TensorRecord(Tensor* _t, unsigned char _ref_count) :
+            t(_t) {}
+};
+
+class TensorShape {
+        TensorShape(uint16_t shape) : _shape({shape, 0, 0, 0}), _num_dims(1) {}
+        TensorShape(uint16_t shape[1]) : _shape({shape, 0, 0, 0}), _num_dims(1) {}
+        TensorShape(uint16_t shape[2]) : _shape({shape[0],shape[1], 0, 0}), _num_dims(2) {}
+        TensorShape(uint16_t shape[3]) : _shape({shape[0], shape[1], shape[2], 0}), _num_dims(3) {}
+        TensorShape(uint16_t shape[4]) : _shape(shape), _num_dims(4) {}
+
+        uint16_t operator[] (int i) const { return _shape[i]; /* Do additional checks*/}
+        uint16_t& operator[](int i) { return _shape[i]; } // Maybe handle update case
+        void update_dims() { 
+            for(int i = 0; i < 4; i++){
+                if(_shape[i] && (i+1) < _num_dims)
+                    _num_dims = i+1;
+                else if(!_shape[i] && (i+1) > _num_dims)
+                    _num_dims = i+1;
+            }
+        }
+        uint16_t get_linear_size() const {
+
+            uint16_t sum = 0;
+            for(int i = 0; i < _num_dims; i++){
+                sum += _shape[i];
+            }
+        }
+    private:
+        uint16_t _shape[4];
+        uint8_t  _num_dims;
+}
+// Do something to remember current type
+enum ttype: uint8_t {
+            i8,
+            u8,
+            i16,
+            u16,
+            i32,
+            u32,
+            float,
+            undefined
+};
+// Need to figure out way of defering reference until after lefthand assignment
+class IntegralValue {
```

Comment:
We need to make sure that the compiler throws this out, verify with disassembly or profiling.

Reply:
It does since most of the class interface is inlineable